### ISSUE TITLE
Identifier `pattern` causes Parse error

### ISSUE THREAD

**Issue by cblp** [ndmitchell/hlint#199]
```haskell
pattern = ()
```
hlint:
```
/tmp/a.hs:1:9: Warning: Parse error: =
Found:
    pattern = ()
    
  > 
  

1 suggestion
```
Looks like `pattern` is recognized as a keyword.

**Comment by ndmitchell** [comment]
Yep, with the new pattern keyword in GHC this code will fail to parse. Have you ought about using a different identifier name? If pattern synonyms ever become default on GHC this code will have issues. Alternatively pass -XNoPatternSynonyms to HLint.

**Comment by cblp** [comment]
In GHC PatterSynonyms extension is disabled by default. Is it good for HLint to have it enabled by default?

**Comment by ndmitchell** [comment]
Yes/no/maybe. I generally enable all extensions that don't break "too many" files. Sometimes people enable their extensions in Cabal or in a build system, so turning on as many as possible makes fewer files fail. In addition, if pattern is becoming a semi-supported GHC keyword, avoiding it in normal use is probably a good idea anyway. There are certain extensions that go on the black list, see https://github.com/ndmitchell/hlint/blob/master/src/Util.hs#L125 - currently:

    badExtensions =
        [Arrows -- steals proc
        ,TransformListComp -- steals the group keyword
        ,XmlSyntax, RegularPatterns -- steals a-b
        ,UnboxedTuples -- breaks (#) lens operator
        ,QuasiQuotes -- breaks [x| ...], making whitespace free list comps break
        ,DoRec, RecursiveDo -- breaks rec
        ]

Given pattern seems like it should be a GHC keyword (given that type, class etc all are), I'd suggest most code rewrites to avoid the word pattern, but if it turns out there's more code than expected, I'll add it.
